Docker Implementation

matthewrsj opened the pull request:
Added Dockerfile and docker_compose.yml for docker functionality. It was also necessary to update several port and host settings. Changes were reflected in the docs.

Commit:
Added host and port variables to conf file for ease of use with docker

Commit:
Updated test form for practice use with docker

Commit:
Added Dockerfile and docker-compose

Commit:
Configured to work in a docker container

Modified Dockerfile to run application only
Modified request_handler to send an email through smtp.osuosl.org
Modified local test form to include 'from' field
Changed assertIsNone tests to use assertEqual(..., None)
  - assertIsNone does not exist in python 2.6, which is what is used
    in the docker container

Commit:
Removed postfix from travis.yml

Commit:
Removed python 2.6

Commit:
Removed personal email from conf

Commit:
Updated docs to reflect docker and settings changes

Commit:
Added one more setting for sendmail

matthewrsj commented on .travis.yml:
```diff
@@ -3,17 +3,6 @@ language: python
-
-notifications:
-  email: false
-
```
Removed all this because it was just covering up a bug that got uncovered by docker.

Basically the sendmail function tried to send an email from localhost but since it wasn't on an osuosl workstation Iit was on Travis) it didn't have the authority to do so. This postfix thing tricked the tests into thinking an email was sent. Fixed by changing the sendmail host to a smtp server.

matthewrsj commented on Dockerfile:
```diff
@@ -0,0 +1,8 @@
+FROM osuosl/python_webapp
```
This is the base image for python webapps from https://github.com/osuosl/osl-dockerfiles/tree/tschuy/python_webapp/python_webapp

matthewrsj commented on Dockerfile:
```diff
@@ -0,0 +1,8 @@
+FROM osuosl/python_webapp
+
+EXPOSE 5000
```
Formsender runs on port 5000 by default, the docker container's port needs to be exposed to communicate with the outside world.

matthewrsj commented on Dockerfile:
```diff
@@ -0,0 +1,8 @@
+
+EXPOSE 5000
+COPY . /opt/formsender
+WORKDIR /opt/formsender
```
Put formsender in /opt/formsender and make that the working directory

matthewrsj commented on conf.py:
```diff
@@ -6,3 +6,6 @@
 DUPLICATE_CHECK_TIME = 3600  # seconds -- 60 seconds * 60 minutes
+HOST = "0.0.0.0"
+PORT = 5000
+SMTP_HOST = "smtp.osuosl.org"
```
Added `HOST` and `PORT` so it is easy to change these in `run_simple()` without trying to find the function call. Explained this in docs.

`SMTP_HOST` is the smtp server `sendmail()` uses. This is what cause so much trouble in Travis.

Kennric replied:
Does this smtp server need authentication of any kind, or can we just send through?

matthewrsj commented on docker-compose.yml:
```diff
@@ -0,0 +1,6 @@
+
+  build: .
+  ports:
+   - "5000:5000"
```
`build: . ` means look locally for the `Dockerfile` to build the container

`- "5000:5000"` binds the exposed docker container port to the local port. The local port can now be used to communicate with Formsender.

matthewrsj commented on request_handler.py:
```diff
@@ -366,9 +367,8 @@ def send_email(msg, email_from, subject):
-    smtp = smtplib.SMTP('localhost')
-    # Attempts to send the mail to EMAIL, with the message formatted as a
-    # string
+    smtp = smtplib.SMTP(SMTP_HOST)
```
This is where the issue was that caused the travis errors and led me down the postfix path...

matthewrsj commented on tests.py:
```diff
@@ -37,7 +37,7 @@ def test_create_msg_no_content(self):
         env = builder.get_environ()
         req = Request(env)
-        self.assertIsNone(handler.create_msg(req))
+        self.assertEquals(handler.create_msg(req), None)
```
My docker image runs python2.6, which doesn't have `assertIsNone()`. After spending too much time trying to update python, I decided to do this. This achieves the same thing.

Kennric commented on docs/source/usage.rst:
```diff
@@ -20,31 +23,34 @@ In conf.py change the following variables to match your setup:
 will email the information submitted to the email specified.
 
 You can change the host and port Formsender waits at by modifying the run_simple
 method at the bottom of request_handler.py. The first argument is the location
```
You don't need to, and shouldn't modify the code now that you have these in config - just explain how to change the config file and remove this part.

Kennric commented on request_handler.py:
```diff
@@ -19,7 +19,8 @@
 from validate_email import validate_email
 from datetime import datetime
-from conf import EMAIL, TOKN, CEILING, DUPLICATE_CHECK_TIME
+from conf import (EMAIL, TOKN, CEILING, DUPLICATE_CHECK_TIME, HOST, PORT,
```
If you are using basically every constant in conf, you might want to just import conf and refer to constant with conf.CONSTANT

Commit:
Made recommended changes

Full import of constants from conf.py
Updated docks to reflect conf.py usefulness

Kennric commented:
This looks good, and it works. +1
